tools/java/org/unicode/cldr/util/XMLSource.java

```diff
@@ -1747,4 +1764,711 @@ public DtdType getDtdType() {
+                case '\'':
+                case '"':
+                    if (inQuote == 0) {
+                        inQuote = ch;
+                    } else if (inQuote == ch) {
+                        inQuote = 0; // come out of quote
+                    }
+                    break;
+                case '/':
+                    if (inQuote == 0 && braceStack == 0) {
+                        return i;
+                    }
+                    break;
+                case '[':
+                    if (inQuote == 0) {
+                        --braceStack;
+                    }
+                    break;
+                case ']':
+                    if (inQuote == 0) {
+                        ++braceStack;
+                    }
+                    break;
+                }
+            }
+            return -1;
+        }
+
+        // SAX items we need to catch
+
+        @Override
+        public void startElement(
+            String uri,
+            String localName,
+            String qName,
+            Attributes attributes)
+            throws SAXException {
+            Log.logln(LOG_PROGRESS || SHOW_START_END, "startElement uri\t" + uri
+                + "\tlocalName " + localName
+                + "\tqName " + qName
+                + "\tattributes " + show(attributes));
+            try {
+                if (isSupplemental < 0) { // set by first element
+                    attributeOrder = new TreeMap<>(
+                        // HACK for ldmlIcu
+                        dtdData.dtdType == DtdType.ldml
+                            ? CLDRFile.getAttributeOrdering()
+                            : dtdData.getAttributeComparator());
+                    isSupplemental = source.dtdType == DtdType.ldml ? 0 : 1;
+                }
+                push(qName, attributes);
+            } catch (RuntimeException e) {
+                e.printStackTrace();
+                throw e;
+            }
+        }
+
+        @Override
+        public void endElement(String uri, String localName, String qName)
+            throws SAXException {
+            Log.logln(LOG_PROGRESS || SHOW_START_END, "endElement uri\t" + uri + "\tlocalName " + localName
+                + "\tqName " + qName);
+            try {
+                pop(qName);
+            } catch (RuntimeException e) {
+                // e.printStackTrace();
+                throw e;
+            }
+        }
+
+        @Override
+        public void characters(char[] ch, int start, int length)
+            throws SAXException {
+            try {
+                String value = new String(ch, start, length);
+                Log.logln(LOG_PROGRESS, "characters:\t" + value);
+                // we will strip leading and trailing line separators in another place.
+                // if (value.indexOf(XML_LINESEPARATOR) >= 0) {
+                // value = value.replace(XML_LINESEPARATOR, '\u0020');
+                // }
+                lastChars += value;
```

Comment:
if `characters` is called a lot then consider making `lastChars` a `StringBuilder`

In general I would search for all instances of `+=`

Reply:
Agree and will change

Reply:
Can't do; this is in the XML parsing code (external).